support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionBuilder.scala

```diff
@@ -5,90 +5,166 @@ import java.util.UUID
+      )
     }
+    else
+      subscriptionData
+  }
 
+  def buildRedemption(
+    today: LocalDate,
+    productRatePlanId: ProductRatePlanId,
+    requestId: UUID,
+    readerType: ReaderType,
+    redemption: SubscriptionRedemption
+  )(implicit ec: ExecutionContext): BuildResult = {
+
+    readerType match {
+      case Corporate => buildCorporateRedemption(
+        today,
+        productRatePlanId,
+        requestId,
+        redemption.redemptionData.redemptionCode,
+        redemption.getCodeStatus
+      )
+      case Gift => buildGiftRedemption(
+        today,
+        productRatePlanId,
+        requestId,
+        redemption.redemptionData.redemptionCode,
+      )
+      case _ => val errorType: Either[PromoError, RedemptionInvalid] = Right(InvalidReaderType)
+        EitherT.leftT(errorType)
+    }
   }
 
-  def withRedemption(
-    subscription: Subscription,
-    redemptionData: RedemptionData,
+  def buildCorporateRedemption(
+    today: LocalDate,
+    productRatePlanId: ProductRatePlanId,
+    requestId: UUID,
+    redemptionCode: RedemptionCode,
     getCodeStatus: GetCodeStatus
-  )(implicit ec: ExecutionContext): EitherT[Future, GetCodeStatus.RedemptionInvalid, Subscription] = {
-    val withCode = subscription.copy(redemptionCode = Some(redemptionData.redemptionCode.value))
-    redemptionData match {
-      case CorporateRedemption(redemptionCode) =>
-        for {
-          subscription <-
-            EitherT(getCodeStatus(redemptionCode).map(_.map { corporateId =>
-              withCode.copy(
-                corporateAccountId = Some(corporateId.corporateIdString),
-                readerType = ReaderType.Corporate
-              )
-            }))
-        } yield subscription
-    }
+  )(implicit ec: ExecutionContext): BuildResult = {
+    val subscriptionData = buildProductSubscription(
+      requestId,
+      productRatePlanId,
+      contractAcceptanceDate = today,
+      contractEffectiveDate = today,
+      readerType = Corporate
+    )
+
+    val redeemedSubcription = for {
+      subscription <-
+        EitherT(getCodeStatus(redemptionCode).map(_.map { corporateId =>
+          subscriptionData.subscription.copy(
+            redemptionCode = Some(redemptionCode.value),
+            corporateAccountId = Some(corporateId.corporateIdString),
+            readerType = ReaderType.Corporate
+          )
+        }))
+    } yield subscription
+
+    redeemedSubcription
+      .map(subscription => subscriptionData.copy(subscription = subscription))
+      .leftMap(Right.apply)
+  }
+
+  def buildGiftRedemption(
+    today: LocalDate,
+    productRatePlanId: ProductRatePlanId,
+    requestId: UUID,
+    redemptionCode: RedemptionCode,
+  )(implicit ec: ExecutionContext): BuildResult = {
+    // TODO: RB implement this
+    val errorType: Either[PromoError, RedemptionInvalid] = Right(InvalidReaderType)
+    EitherT.leftT(errorType)
   }
 }
```

Comment:
probably worth splitting this file/object further into redemption and purchase related stuff.  This would let us have a more clear grouping about which behaviours apply to each situation and which are common to both.

Reply:
I'll address this more in the next PR